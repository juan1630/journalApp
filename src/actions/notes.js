import Swal from 'sweetalert2';

import { db } from "../firebase/firebase-config";
import { fileUploas } from '../helpers/fileUpload';
import { loadNotes } from "../helpers/loadNotes";
import { types } from '../types/types'

export const startNewNote = ( ) => {
    
    // el getState da accesio a todo el state
    return  async (dispatch, getState ) => {

        // obtenemos el UID del usuario
        const {uid} = getState().auth;

        // inicializmaos el objeto del note
        const newNote = {
            title: '',
            body: '',
            dates: new Date().getTime()
        }

        // sirve para en viar a la DB los datos y grbarlos en la rura que se especifica
        const docRef = await db.collection(`${uid}/journal/notes`).add(   newNote );
        // console.log(docRef);

        // hacemos el dispartch de la funcio ativatesNotes que se encuntra abajo
        dispatch( activeNotes( docRef.id, newNote ) );
        dispatch(  addNewNote(docRef.id, newNote) );
    }
}



export const activeNotes = (id, note) => ({
 
    type: types.notesActive,
    payload : {
        id,
        ...note
    }
});
 export const addNewNote = (id, note) => ({
     type:types.notesAddNew,
     payload : {
        id, ...note
     }
 })

export const startLoadNotes = ( uid ) => {

    return async ( dispatch ) => {
        
        const notes = await loadNotes(  uid );
        dispatch( setNotes( notes));

    }

}

export const setNotes = ( notes ) => ({

    type: types.notesLoad,
    payload : notes
    
})


export const startSavedNote = (note) => {
    
    return async (dispatch, getState) => {

        console.log(note);

        const { uid } = getState().auth;


        if(!note.url ) {
            delete note.url;
            // borramos esta propiedad del json
            delete note.date;
        }   
        const noteToFirestore = { ...note }
        console.log(noteToFirestore)
        
        // eliminamos el id del body
        delete noteToFirestore.id;
        delete noteToFirestore.date;


        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

        dispatch( refreshNotes( note.id, note ) );
        Swal.fire('Save', note.title, 'success');

    }
}

export const refreshNotes = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})




export const startFileUpload  = (file) => {
    
    return async ( dispatch,  getState ) => {
        
        const { active: activeNotes } = getState().notes;
        // console.log(file, activeNotes)
        const fileUrl = await fileUploas(file);
        // console.log( fileUrl );
        Swal.fire({
            title: 'Uploading',
            text: 'Please upload',
            allowOutsideClick: false,
            onBeforeUpload: () =>{
                Swal.showLoading();
            }
        });

        activeNotes.url = fileUrl;
        Swal.close();


        dispatch( startSavedNote(activeNotes) )

    }
}




export const startDeleting = (id) => {
    
    return async ( dispatch, getState ) => {
        const { uid }  = getState().auth;
        console.log( `${uid}/journal/notes/${id}`)
        await db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch( deleteNote(id) );
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
});


export const noteLogout = () => ({

    type: types.notesLogoutCleaning,

});
