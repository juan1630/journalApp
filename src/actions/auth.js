// imports de otras despencias 
import Swal from 'sweetalert2';

import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config'
import { startLoading, finishLoading } from './ui'
import { noteLogout } from './notes';


export const loginWithEmailYPassword = (email, password) => {
    
    // esta funcion retorna un callback
    return (dispatch) => {

        dispatch( startLoading() );

        return firebase.auth().signInWithEmailAndPassword(email, password)
        .then(({user}) => {

            dispatch( finishLoading() );
            dispatch( login( user.uid, user.displayName ) );
        })
        .catch( e => {
            Swal.fire('Fail', e.message, 'error');
            dispatch(finishLoading() )
        });
    }
}


export const startGoogleLgoin = () => {
    
    return (dispatch) => {

        //hacemos uso de la funcio auth y el provider es google que esta declarado en el config de firebase
        firebase.auth().signInWithPopup( googleAuthProvider ) 
        .then( ({user})  => {

            dispatch( login( user.uid, user.displayName ));
        })
        .catch( (err) => {
            Swal.fire('Fail', err.message, 'error');
        });
    }
}


export const starRegisterWithEmailPasswordEmail = (name, email, password) => {

    return (dispatch) => {
         firebase.auth().createUserWithEmailAndPassword(email, password)
         .then(  async ({user}) => {

            // actualizamos el displayName porque viene en nulo cuando se hace el registro
          await  user.updateProfile({displayName: name})
                dispatch( login( user.uid, user.displayName) );
         })
         .catch( error  => console.error(error))
    }
}




export const startLogout = () => {
    return async (dispatch) => {
        // esta funcion regresa una promesa
        await firebase.auth().signOut();
        
        dispatch(logOut());
        dispatch( noteLogout() );
    }
}



export const login = (uid, displayName) =>  ({
        type: types.loggin,
        payload: {
            uid,
            displayName,
        }
});

export const logOut = () =>({
    type: types.logout
})