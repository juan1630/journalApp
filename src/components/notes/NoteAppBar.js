import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startSavedNote } from '../../actions/notes';

export const NoteAppBar = () => {
    
    const dispatch = useDispatch();
    const { active } = useSelector( state => state.notes );

    const handleSaveNote = () => {
        console.log(active);
        dispatch( startSavedNote( active ) );
    }

    return (
        <div className="notes__appbar" >
            <span> 28 de agosto de 2021</span>
            <div>
                <button className="btn"> Picture </button>    
            </div>
            <div>
                <button 
                    className="btn"
                    onClick={   handleSaveNote }            
                    > Save </button>    
            </div>
        </div>
    )
}
