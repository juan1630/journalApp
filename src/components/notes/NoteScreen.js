import React, { useEffect, useRef} from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { NoteAppBar } from './NoteAppBar';
import { useForm } from '../../hooks/useForm';
import { activeNotes, startDeleting } from '../../actions/notes';
// import { useRef } from 'react';

export const NoteScreen = () => {

        const { active: note } = useSelector(state => state.notes);

        const dispatch = useDispatch();
        // Renombemos la varibale por note
        
        const [ formValues, handleInputChange, reset ] = useForm(note);
        // incializamos el form con los valores del note

        const activeId = useRef(note.id);
        // este hook hace referencia a una variable que no redibuja el DOM cuando el valor cambia
        // console.log(note);

        const { body, title} = formValues;

        useEffect(() => {  
            
            // con el currente entramos al valor actual de la variable
            if(note.id !== activeId.current) {
                reset(note);
                activeId.current = note.id;
            }
         }, [note, reset]);

         useEffect(() => {
            dispatch( activeNotes(formValues.id, {
                ...formValues,
            }) )
         }, [formValues, dispatch]);

         const handleDelete = () => {
             dispatch(  startDeleting( note.id) )
         }

    return (
        <div  className="note-screen-content">
            
            <NoteAppBar />

            <div className="notes__content" >
                <input type="text" 
                        className="notes__title-input"
                        placeholder="an aweson titlfor everyone"
                        autoComplete="off"
                        value={title}
                        name="title"
                        onChange={handleInputChange}
                        />
                <textarea 
                    placeholder="What happend today"
                    className="notes__textarea"
                    value={ body }
                    name="body"
                    onChange={handleInputChange}
                    >

                </textarea>
            </div>
               
                {
                    (note.url) && 
                    (   <div  className="notes__image" >
                            <img src={note.url} alt="an im" />
                        </div>
                    )
                }


        <button 
            className="btn btn-danger"
            onClick={ handleDelete }
            > 
            Delete  
        </button>

        </div>
    )
}
