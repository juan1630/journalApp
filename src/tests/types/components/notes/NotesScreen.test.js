
import React from 'react';
import {mount} from 'enzyme';


import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk'; 
import '@testing-library/jest-dom';

// permite finguir la rutas en el ambiente de testing

import { Provider } from 'react-redux';

import { NoteScreen } from '../../../../components/notes/NoteScreen';
import { activeNotes } from '../../../../actions/notes';



// jest.mock('../../../actions/auth', () => ({
//     startGoogleLgoin: jest.fn(),
// }));z

jest.mock('../../../../actions/notes', () => ({
    activeNotes: jest.fn(),
}));

// creamos un mock de la funcion que vamos a llamar




const middlewares = [ thunk ];

// mocks 
const mockStore = configureStore(middlewares);
const initState = {
    auth: {},
    uid: {
        loading: false,
        msgError: null
    }, 
    notes : {
            active: {
                id:123,
                title: '',
                body : ''
            }
    }
};

let store = mockStore(initState);
store.dispatch = jest.fn();


const wrapper = mount(
                <Provider store={store} >

                    <NoteScreen/> 
                </Provider>
                 
                 );

describe('Pruebas en el NotesScreen', () => {
    
    test('Debe de mostarse correctamente', () => {
        
        expect(wrapper).toMatchSnapshot();

    });

    test('Debe de disparar el action note', () => {

        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value : 'Hola de nuevo'
            }
        });

    });


    test('Debe de llamarse el note', () => {
        // manda a llamar la ultina funcionn
        expect( activeNotes ).toHaveBeenLastCalledWith(
            123,
            {
                title: 'Hola de nuevo',
                body : '',
                id:123
            }
        );
    });
    
    
});