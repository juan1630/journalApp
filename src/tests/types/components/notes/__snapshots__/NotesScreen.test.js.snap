// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el NotesScreen Debe de mostarse correctamente 1`] = `
<div
  className="note-screen-content"
>
  <div
    className="notes__appbar"
  >
    <span>
       28 de agosto de 2021
    </span>
    <input
      id="fileSelector"
      name="fileInput"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
         Picture 
      </button>
    </div>
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
         Save 
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="an aweson titlfor everyone"
      type="text"
      value=""
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happend today"
      value=""
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
